### PR TITLE
Test modeling rules junit

### PR DESCRIPTION
**Pull request by kobymeir**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-6879
SDK https://github.com/demisto/demisto-sdk/pull/3475

## Description
Test modeling rules junit

### CODE REVIEW

File: .gitlab/ci/.gitlab-ci.on-push.yml [hunk 1]
```diff
@@ -546,6 +544,14 @@ xsiam_server_ga:
...
     MARKETPLACE_NAME: "marketplacev2"
     UNREMOVABLE_PACKS: "Base"
+    JIRA_COMPONENT: "Test Modeling Failure"
+    JIRA_ISSUE_TYPE: "XSOAR Dev Bug"
+    JIRA_ISSUE_UNRESOLVED_TRANSITION_NAME: "Backlog"
+    JIRA_PROJECT_ID: "CIAC"
+    JIRA_SERVER_URL: "https://jira-hq.paloaltonetworks.local"
+    JIRA_VERIFY_SSL: "False"
+    JIRA_LABELS: '["XSIAM"]'
+    JIRA_ADDITIONAL_FIELDS: '{"customfield_17166":{"value": "IN"},"customfield_17167":{"value": "FALSE"},"customfield_17168":{"value": "Single"}}'
```
Comment (DeanArbel) [inline .gitlab/ci/.gitlab-ci.on-push.yml:554]: Please add a comment about what are the custom fields provided here.

File: Tests/scripts/convert_test_result_to_jira_issues.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        *{properties['pack_id']}* - *{properties['file_name']}* failed in {build_markdown_link}
+
+        ||Tests||Result||
+        |Total|{test_suite.tests}|
+        |Failed|{test_suite.failures}|
+        |Errors|{test_suite.errors}|
+        |Skipped|{test_suite.skipped}|
+        |Successful|{test_suite.tests - test_suite.failures - test_suite.errors - test_suite.skipped}|
+        |Duration|{test_suite.time}s|
+        """
```
Comment (DeanArbel) [inline Tests/scripts/convert_test_result_to_jira_issues.py:64]: Please extract the description creation to its own function.

File: Tests/scripts/convert_test_result_to_jira_issues.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                                  f"transition for issue {searched_issue.key}")
+                    jira_issue = None
+                    use_existing_issue = False
+                    link_to_issue = searched_issue
+
+            else:
+                link_to_issue = searched_issue
+        else:
+            jira_issue = searched_issue
+
```
Comment (DeanArbel) [inline Tests/scripts/convert_test_result_to_jira_issues.py:102]: Please consider extracting this to its own function.

File: Tests/scripts/install_content_and_test_integrations.sh [hunk 4]
```diff
@@ -15,7 +15,17 @@ if [ -n "${NIGHTLY}" ]; then
   IS_NIGHTLY=true
 fi
 
-python3 ./Tests/configure_and_test_integration_instances.py -u "$USERNAME" -p "$PASSWORD" -c "$CONF_PATH" -s "$SECRET_CONF_PATH" --tests_to_run "$ARTIFACTS_FOLDER/filter_file.txt"  --pack_ids_to_install "$ARTIFACTS_FOLDER/content_packs_to_install.txt" -g "$GIT_SHA1" --ami_env "$1" -n $IS_NIGHTLY --branch "$CI_COMMIT_BRANCH" --build-number "$CI_PIPELINE_ID" -sa "$GCS_MARKET_KEY" --build_object_type "$2" --cloud_machine "$CLOUD_CHOSEN_MACHINE_ID" --cloud_servers_path "$CLOUD_SERVERS_PATH" --cloud_servers_api_keys "cloud_api_keys.json" --marketplace_name "$MARKETPLACE_NAME" --artifacts_folder "$ARTIFACTS_FOLDER" --marketplace_buckets "$GCS_MACHINES_BUCKET"
 if [ -f ./Tests/test_pack.zip ]; then
+  echo "Copying test_pack.zip to artifacts folder:${ARTIFACTS_FOLDER}"
```
Comment (DeanArbel) [inline Tests/scripts/install_content_and_test_integrations.sh:19]: How come this step now happens before?

File: .gitlab/ci/.gitlab-ci.on-push.yml [hunk 5]
```diff
@@ -532,6 +522,14 @@ xsiam_server_ga:
     - if: '$CI_PIPELINE_SOURCE =~ /^(push|contrib)$/'
     - if: '$NIGHTLY'
       when: always
+  artifacts:
+    when: always
+    expire_in: 30 days
```
Comment (tkatzir) [inline .gitlab/ci/.gitlab-ci.on-push.yml:527]: 30 days sounds a lot.

File: Tests/scripts/convert_test_result_to_jira_issues.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+from junitparser import TestSuite, JUnitXml
+from tabulate import tabulate
+
+from Tests.scripts.utils import logging_wrapper as logging
+from Tests.scripts.utils.log_util import install_logging
+
+urllib3.disable_warnings()  # Disable insecure warnings
+
+GITLAB_PROJECT_ID = os.getenv('CI_PROJECT_ID') or 2596  # the default is the id of the content repo in code.pan.run
+GITLAB_SERVER_URL = os.getenv('CI_SERVER_URL', 'https://code.pan.run')  # disable-secrets-detection
```
Comment (tkatzir) [inline Tests/scripts/convert_test_result_to_jira_issues.py:20]: Do we need to expose these internal values?

File: Tests/scripts/convert_test_result_to_jira_issues.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+# Jira label are a json string that will be parsed into a list of labels.
+JIRA_LABELS = json.loads(os.environ.get("JIRA_LABELS", "[]"))
+
+
+def options_handler() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description='Converts the junit report to a jira issues')
+    parser.add_argument("-jp", "--junit-path", help='JUnit report file path', required=True)
+    parser.add_argument('-u', '--url', help='The gitlab server url', default=GITLAB_SERVER_URL)
+    parser.add_argument('-gp', '--gitlab_project_id', help='The gitlab project id', default=GITLAB_PROJECT_ID)
+    parser.add_argument('-d', '--max_days_to_reopen', default=30, type=int,
```
Comment (tkatzir) [inline Tests/scripts/convert_test_result_to_jira_issues.py:41]: Some arguments are kebab-case, and some are snake_case,

File: Tests/scripts/convert_test_result_to_jira_issues.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        logging.info(f'Gitlab server url: {options.url}')
+        logging.info(f'Gitlab project id: {options.gitlab_project_id}')
+        logging.info(f'Jira server url: {JIRA_SERVER_URL}')
+        logging.info(f'Jira verify SSL: {JIRA_VERIFY_SSL}')
+        logging.info(f'Jira project id: {JIRA_PROJECT_ID}')
+        logging.info(f'Jira issue type: {JIRA_ISSUE_TYPE}')
+        logging.info(f'Jira component: {JIRA_COMPONENT}')
+        logging.info(f'Jira labels: {JIRA_LABELS}')
+        logging.info(f'Jira issue unresolved transition name: {JIRA_ISSUE_UNRESOLVED_TRANSITION_NAME}')
+        logging.info(f'Max days to reopen: {options.max_days_to_reopen}')
```
Comment (tkatzir) [inline Tests/scripts/convert_test_result_to_jira_issues.py:165]: Can they be consolidated into one log message?

File: Tests/scripts/convert_test_result_to_jira_issues.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+        jira_server = JIRA(JIRA_SERVER_URL, token_auth=JIRA_API_KEY, options={'verify': JIRA_VERIFY_SSL})
+
+        xml = JUnitXml.fromfile(options.junit_path)
+        for test_suite in xml.iterchildren(TestSuite):
+            if test_suite.failures or test_suite.errors:
+                create_jira_issue(jira_server, test_suite, options, now)
+            else:
+                logging.info(f"Skipped creating Jira issue for {test_suite.name}")
```
Comment (tkatzir) [inline Tests/scripts/convert_test_result_to_jira_issues.py:174]: Maybe this should be debug level?
```suggestion
                logging.info(f"Skipped creating Jira issue for successful test {test_suite.name}")
```

File: Tests/scripts/gitlab_slack_notifier.py [hunk 10]
```diff
@@ -134,23 +167,22 @@ def bucket_upload_results(bucket_artifact_folder, should_include_private_packs:
...
     if failed_packs:
         steps_fields += [{
             'title': f'Failed {marketplace_name} Packs:',
-            'value': ', '.join(sorted([pack_name for pack_name in {*failed_packs}], key=lambda s: s.lower())),
+            'value': ', '.join(sorted({*failed_packs}, key=lambda s: s.lower())),
             'short': False
         }]
 
     if successful_private_packs and should_include_private_packs:
         # No need to indicate the marketplace name as private packs only upload to xsoar marketplace.
```
Comment (tkatzir) [inline Tests/scripts/gitlab_slack_notifier.py:182]: Is there a distinction between the XSOAR and XSOAR-NG/XSOAR-SAAS marketplaces?
If yes, then maybe we need to indicate the marketplace here?

File: Tests/scripts/test_modeling_rules.sh [hunk 11]
```diff
@@ -24,10 +23,34 @@ for modeling_rule in "${MODELING_RULES_ARRAY[@]}"; do
...
+  echo "Testing Modeling Rules - demisto-sdk version is too old, skipping junit report generation"
+  MODELING_RULES_RESULTS_ARG=()
+  MODELING_RULES_RESULTS_FILE_NAME=""
+fi
+
+demisto-sdk modeling-rules test --xsiam-url="$XSIAM_URL" --auth-id="$AUTH_ID" --api-key="$API_KEY" \
+  --xsiam-token="$XSIAM_TOKEN" --non-interactive "${MODELING_RULES_RESULTS_ARG[@]}" \
+  ${MODELING_RULES_TO_TEST}
+exit_code=$?
+
```
Comment (tkatzir) [inline Tests/scripts/test_modeling_rules.sh:45]: Or add it to an `else` of the `if` below.
```suggestion
  echo "Finished testing Modeling Rules, exit_code=$exit_code"

```

File: pyproject.toml [hunk 12]
```diff
@@ -10,7 +10,8 @@ python = "^3.8,<3.11"
 defusedxml = "^0.7.1"
 
 [tool.poetry.group.dev.dependencies]
-demisto-sdk = "1.19.1"
+#demisto-sdk = "1.19.1"
+demisto-sdk = { git = "https://github.com/demisto/demisto-sdk.git", rev = "test-modeling-rules-junit" }
```
Comment (tkatzir) [inline pyproject.toml:14]: Remove before merging

**Review by tkatzir — APPROVED** [review]
Very good!
Approved, with some comments (Some are important before merging...)

File: Tests/scripts/convert_test_result_to_jira_issues.py (hunk 7 above)
Comment (kobymeir) [inline Tests/scripts/convert_test_result_to_jira_issues.py:41]:
```suggestion
    parser.add_argument('-d', '--max_days_to_reopen', default=30, type=int,  # fixme kebab-case all args
```

File: Tests/scripts/convert_test_result_to_jira_issues.py (hunk 6 above)
Comment (kobymeir) [reply Tests/scripts/convert_test_result_to_jira_issues.py:20]: it's exposed in all other scripts, i will remove it.

File: Tests/scripts/convert_test_result_to_jira_issues.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+import urllib3
+from jira.client import ResultList, Issue, JIRA
+from junitparser import TestSuite, JUnitXml
+from tabulate import tabulate
+
+from Tests.scripts.utils import logging_wrapper as logging
+from Tests.scripts.utils.log_util import install_logging
+
+urllib3.disable_warnings()  # Disable insecure warnings
+
```
Comment (kobymeir) [inline Tests/scripts/convert_test_result_to_jira_issues.py:18]:
```suggestion
  # fixme remove default values
```

File: Tests/scripts/convert_test_result_to_jira_issues.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        """
+    return description
+
+
+def main():
+    try:
+        install_logging('convert_test_result_to_jira_issues.log', logger=logging)
+        now = datetime.now(tz=timezone.utc)
+        options = options_handler()
+        logging.info(f'JUnit path: {options.junit_path}')
```
Comment (kobymeir) [inline Tests/scripts/convert_test_result_to_jira_issues.py:155]:
```suggestion
        logging.info(f'JUnit path: {options.junit_path}'). # fixme join logger.info
```

File: Tests/scripts/gitlab_slack_notifier.py (hunk 10 above)
Comment (kobymeir) [reply Tests/scripts/gitlab_slack_notifier.py:182]: @tkatzir read the comment on line 149:

`        # No need to indicate the marketplace name as private packs only upload to xsoar marketplace.`

File: .gitlab/ci/.gitlab-ci.on-push.yml (hunk 5 above)
Comment (kobymeir) [inline .gitlab/ci/.gitlab-ci.on-push.yml:527]:
```suggestion
    expire_in: 14 days
```

File: .gitlab/ci/.gitlab-ci.yml [hunk 15]
```diff
@@ -46,13 +46,13 @@ variables:
   GCS_PRODUCTION_V2_BUCKET: "marketplace-v2-dist"
   GCS_PRODUCTION_XPANSE_BUCKET: "xpanse-dist"
   DEMISTO_CONNECTION_POOL_MAXSIZE: "180" # see this issue for more info https://github.com/demisto/etc/issues/36886
-  SDK_REF: "master" # The default sdk branch to use
+  SDK_REF: "test-modeling-rules-junit" # FIXME The default sdk branch to use
```
Comment (kobymeir) [inline .gitlab/ci/.gitlab-ci.yml:49]:
```suggestion
  SDK_REF: "master"
```

File: .gitlab/ci/.gitlab-ci.yml [hunk 16]
```diff
@@ -46,13 +46,13 @@ variables:
...
   OVERRIDE_ALL_PACKS: "false"
   TEST_UPLOAD: "true"
   NATIVE_CANDIDATE_IMAGE: "latest"
   DEMISTO_SDK_GRAPH_FORCE_CREATE: "true"  # change this when the demisto-sdk update-graph command is stable
   DEMISTO_SDK_LOG_FILE_PATH: $ARTIFACTS_FOLDER/logs/demisto_sdk_debug.log
-  OVERRIDE_SDK_REF: ""
+  OVERRIDE_SDK_REF: "1" # FIXME revert to empty.
```
Comment (kobymeir) [inline .gitlab/ci/.gitlab-ci.yml:55]:
```suggestion
  OVERRIDE_SDK_REF: ""
```

File: .gitlab/ci/.gitlab-ci.yml [hunk 17]
```diff
@@ -46,7 +46,7 @@ variables:
   GCS_PRODUCTION_V2_BUCKET: "marketplace-v2-dist"
   GCS_PRODUCTION_XPANSE_BUCKET: "xpanse-dist"
   DEMISTO_CONNECTION_POOL_MAXSIZE: "180" # see this issue for more info https://github.com/demisto/etc/issues/36886
-  SDK_REF: "master" # The default sdk branch to use
+  SDK_REF: "master"
```
Comment (kobymeir) [inline .gitlab/ci/.gitlab-ci.yml:49]:
```suggestion
  SDK_REF: "master" # The default sdk branch to use
```

**Pull request merged**